components/tabs/src/sync/record.rs

```diff
@@ -27,7 +29,379 @@ pub struct TabsRecord {
+                let title: String = parse_string_from_json(&"title", tabs_obj.get("title"))?;
+                let url_history: Vec<String> = tabs_obj
+                    .get("urlHistory")
+                    .ok_or_else(|| serde_json::Error::custom("missing field `urlHistory`"))?
+                    .as_array()
+                    .ok_or_else(|| {
+                        serde_json::Error::custom("invalid `urlHistory`, expected sequence")
+                    })?
+                    .iter()
+                    .map(|u| -> Result<String> {
+                        Ok(u.as_str()
+                            .ok_or_else(|| {
+                                serde_json::Error::custom(
+                                    "invalid `urlHistory` value, expected string",
+                                )
+                            })?
+                            .to_string())
+                    })
+                    .into_iter()
+                    .collect::<Result<_>>()?;
+                let icon: Result<Option<String>, serde_json::Error> =
+                    if let Some(icon) = tabs_obj.get("icon") {
+                        if let Some(i) = icon.as_str() {
+                            Ok(Some(i.to_string()))
+                        } else {
+                            Err(serde_json::Error::custom("invalid `icon`, expected string"))
+                        }
+                    } else {
+                        Ok(None)
+                    };
+
+                let last_used = parse_last_used(
+                    tabs_obj
+                        .get("lastUsed")
+                        .ok_or_else(|| serde_json::Error::custom("missing field `lastUsed`"))?,
+                )?;
+
+                Ok(TabsRecordTab {
+                    title,
+                    url_history,
+                    icon: icon?,
+                    last_used,
+                })
+            })
+            .into_iter()
+            .collect::<Result<_>>();
+
+        let ttl: u32 = payload.data["ttl"].as_u64().unwrap_or_default() as u32;
+
+        Ok(TabsRecord {
+            id: payload.id.to_string(),
+            client_name,
+            tabs: tabs?,
+            ttl,
+        })
+    }
+}
+
+fn parse_last_used(last_used_val: &JsonValue) -> Result<u64> {
+    // In order to support older clients, we are handling `last used` values that
+    // are either floats, integers, or stringified floats or integers. We attempt to
+    // represent `last used` as a float before converting it to an integer. If that
+    // operation isn't successful, we try converting `last used` to an integer directly.
+    // If that isn't successful, the returned value will be zero.
+
+    let invalid_err_msg = "invalid `lastUsed`, expected u64";
+    let last_used = if last_used_val.is_string() {
+        let l = last_used_val.as_str().unwrap_or_default();
+
+        match l.parse::<f64>() {
```

Comment:
(This is a comment on Rust, not on your code)

I simultaneously love and hate this `parse` syntax...like, it's so powerful and so clever and at the same time, so darn *weird* for what seems like a really basic operation 😬. Anyways...